Ak96/rewrite in python attempt 2

aaruni96 opened the pull request:
A better attempt of rewriting the update stuff from bash into python.

This requires PyGithub to be installed on the server.

The API_KEY is defined in the systemd unit file for the website update : `~/.config/systemd/user/oscar-website.service`

When a network failure occurs, the behavior is different based on what we are updating:

- Tutorial status: keeps file unchanged
- Tutorial last updated date: uses fallback date (01-01-1970) for the tutorials it couldn't fetch date for, new date for those it could fetch
- OSCAR release status: keeps file unchanged

aaruni96 committed:
Reapply "Rewrite in python / add fallback for network failures"

This reverts commit f0d94d3de2e4b25ba269e23a4b8a38c6371cbd68.

aaruni96 committed:
More logs and better failure fallbacks for update scripts

aaruni96 committed:
Don't return exit code, or main bash script will stop

fingolfin commented on etc/update.sh:
```diff
@@ -15,24 +15,17 @@ bundle config set --local path 'vendor/bundle'
-statusurl=$(curl -SsL --retry 5 "https://api.github.com/repos/oscar-system/TutorialTesterforOscar/actions/workflows/CI.yml/runs?per_page=1" | jq  '.workflow_runs[0].jobs_url' | sed 's/"//g')
-curl -SsL --retry 5 $statusurl | jq '.["jobs"][1:] | .[] | .name+":"+.conclusion' | sed 's/^.*\s.*\s.*\s//' | sed 's/):/: /' | sed s'/"$//' > _data/examples_status.yml
+echo "Getting tutorial status....."
+./etc/tutorial_status.py
```
You are now executing a Python script that has prerequisites. So this `update.sh` script should ensure those are actually satisfied / present. I.e. we need the analogue to `bundle install` above.

For GAP we do this in our CI when we need to run a Python script:
```
python -m pip install -r dev/releases/requirements.txt
```
where `dev/releases/requirements.txt` is plain text file with each line a package name.

Perhaps it should also pass `--user` to `pip` and / or should be setting up / use a Python venv .

aaruni96 replied:
I have added a venv activation phase to the script, but this will require us to have a one time creation of the venv manually before this PR is merged.

fingolfin replied:
As suggested elsewhere, I don't see why you can't also init the venv here if it is not yet there? Or even just init it unconditionally?

fingolfin commented on etc/update.sh:
```diff
@@ -15,24 +15,17 @@ bundle config set --local path 'vendor/bundle'
 # run jekyll
-bundle exec jekyll build --config _config.yml,_config_production.yml -d /srv/www/www-mathe-oscar/data/http
+echo "Running jekyl......."
+bundle exec jekyll build --config _config.yml,_config_production.yml # -d /srv/www/www-mathe-oscar/data/http
```
why this change? is it intentional?

aaruni96 replied:
No, something carried forward from the last PR and I should've caught.

fingolfin commented on etc/update-latest-release.py:
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python3
+
+import os
```
```suggestion
import os
import sys
```

fingolfin commented on etc/update-latest-release.py:
```diff
@@ -0,0 +1,41 @@
+
+API_KEY = os.getenv("API_KEY")
+if API_KEY == None:
+    print("API key was not found! Authentication will fail!\nSet the environment variable API_KEY "
+          "to a github access token and try again!\n")
+
+assert API_KEY != None
```
It seems weird to catch an error, print a nice error message, but then proceed and instead rely on an `assert` to throw an error?
```suggestion
    print("API key was not found! Authentication will fail!\nSet the environment variable API_KEY "
          "to a github access token and try again!\n")
    sys.exit(1)
```

aaruni96 replied:
`sys.exit(1)` (and also a failing assert) will stop the update.sh script.

Should we use the builtin `exit()`, so that we exit from the python script, but the update script keeps going ?

fingolfin replied:
Fine by me

fingolfin commented on etc/update-latest-release.py:
```diff
@@ -0,0 +1,41 @@
+    print(e)
+    print("Network access failed!")
+    print("Leaving the release file unchanged!")
+    os.exit()
```
I don't think there is an `os.exit`. Did you perhaps mean this?
```suggestion
    sys.exit()
```

aaruni96 replied:
I meant to use the builtin exit function, but my brain managed to convince me in the moment that I should just replace `sys` to `os`

fingolfin commented on etc/update-latest-release.py:
```diff
@@ -0,0 +1,41 @@
+if os.getcwd().split('/')[-1] == 'etc':
+    releasefilepath = '../_data/release.yml'
+else:
+    releasefilepath = "_data/release.yml"
```
This is very brittle. I would suggest we either require this path as an argument to the script; or else let the script locate itself (`os.path.abspath(sys.argv[0])`) and then take a path relative to that.

aaruni96 replied:
Does the commit https://github.com/oscar-system/oscar-website/pull/503/commits/9ce5ba3cd202f5e248a4c08d7f1bee96637c67f7 properly implement what you suggested ?

fingolfin replied:
it looks plausible. but of course it should be tested :-)

aaruni96 committed:
Fix a problem carried over by reverting the revert

aaruni96 committed:
Exit in the same block as when we check for API_KEY

We are using inbuilt exit() instead of assert or sys.exit(1) so that
this does not stop the update.sh script

aaruni96 committed:
Less brittle way of finding the data directory for script output

aaruni96 committed:
Add sourcing from venv, and requirements.txt

fingolfin commented on etc/update.sh:
```diff
@@ -14,25 +14,27 @@ cat /srv/www/www-mathe-oscar/data/webhook.secret >> .htaccess
 
+# use the venv defined in .venv
+# set this up before running this script!
+echo "Activating venv..."
```
can't you add a line before this doing roughly `if .venv is not an existing dir, set up a venv` ?

That will make it that much easier to set up a new web server if we ever have to. The alternative would be to add this as another step in `etc/README.md` (I think adding it to the script is preferable to that)

fingolfin commented on etc/tutorial_status.py:
```diff
@@ -0,0 +1,50 @@
+    print("Done!")
+except Exception as e:
+    print(e)
+    falied = True
```
```suggestion
    failed = True
```

fingolfin commented on etc/tutorial_status.py:
```diff
@@ -0,0 +1,50 @@
+datapath = '/'.join(os.path.abspath(sys.argv[0]).split('/')[0:-2])+'/_data'
+statusfilepath = f"{datapath}/examples_status.yml"
+
+if resultstring=="":
+    # if resultstring is empty, just use the old data
```
but we are not using the old state, are we? we are overwriting it now with the content of `example_status.yml` which may be old ?

Moreover, if `resultstring` is *not* empty, we are now discarding it, that seems wrong?

So maybe
```suggestion
if len(resultstring) > 0:
```


But once you do all this, couldn't you remove the `try/except`, the variable `failed` etc. -- and then the script just aborts (doing nothing) if there is an error.

In in `update.sh`, just ignore the exit code

aaruni96 replied:
Sorry, this was supposed to be `resultstring!=""` (and your suggestion of checking length is even better).

I think the idea of `try/except` was to get a good error message. But thinking about it.., we can keep the `try/except`, and move the pretty error message to the `except` part, and then exit. This would let us get rid of the `failed` variable.

aaruni96 replied:
Does commit 7713298d look alright ?

fingolfin replied:
It looks plausible.

but I don't underestimate why you then still need to check if resultstring is non-empty?

aaruni96 replied:
To guard against my paranoia for something that has very little chance of happening :

What we do in this script, is fetch the job details of the latest CI run of the tutorial tester repository ( https://github.com/oscar-system/TutorialTesterforOscar/ ). The CI run of tutorial tester fetches a job list dynamically, and it is possible that something goes wrong there, and it returns an empty list. In this case, the CI job would appear to run successfully with 0 entries. So, when we fetch this particular job, the `try/except` will succeed, and resultstring will still be empty.

I know, very unlikely to happen, but the possibility of this happening popped into my head while I was writing the script.

fingolfin commented on etc/update.sh:
```diff
@@ -14,25 +14,27 @@ cat /srv/www/www-mathe-oscar/data/webhook.secret >> .htaccess
-statusurl=$(curl -SsL --retry 5 "https://api.github.com/repos/oscar-system/TutorialTesterforOscar/actions/workflows/CI.yml/runs?per_page=1" | jq  '.workflow_runs[0].jobs_url' | sed 's/"//g')
-curl -SsL --retry 5 $statusurl | jq '.["jobs"][1:] | .[] | .name+":"+.conclusion' | sed 's/^.*\s.*\s.*\s//' | sed 's/):/: /' | sed s'/"$//' > _data/examples_status.yml
+echo "Getting tutorial status....."
+./etc/tutorial_status.py
```
To ignore the exit code, could try this:
```suggestion
./etc/tutorial_status.py || :
```

fingolfin commented on etc/update-latest-release.py:
```diff
@@ -0,0 +1,41 @@
+    print("Leaving the release file unchanged!")
+    exit()
+
+if not failed:
```
I think `failed` is redundant now in this script and could be removed.

aaruni96 committed:
automate initializing the venv

aaruni96 committed:
Remove failed variable, move erorr message to except block

fingolfin commented on etc/tutorial_status.py:
```diff
@@ -0,0 +1,47 @@
+    print(f"Processing {i.name}.....")
+    name = i.name.split()[-1][0:-1]
+    status = i.conclusion
+    resultstring += f"{name}: {status}\n"
```
For extra resilience I would suggest to generate YAML using quoted key/value pairs.
```suggestion
    resultstring += f"'{name}': '{status}'\n"
```

fingolfin commented on etc/update.sh:
```diff
@@ -14,25 +14,31 @@ cat /srv/www/www-mathe-oscar/data/webhook.secret >> .htaccess
+./etc/update-latest-release.py
+echo "Done!"
 # run jekyll
+echo "Running jekyl......."
```
```suggestion
echo "Running jekyll......."
```
Instead of all these manuals print statements it may be more effective to add `set -x` at the start of the script

aaruni96 committed:
Quote yaml key/value pair

Suggested in review

Co-authored-by: Max Horn <max@quendi.de>

fingolfin commented on etc/tutorial_status.py:
```diff
@@ -0,0 +1,47 @@
+    status = i.conclusion
+    resultstring += f"'{name}': '{status}'\n"
+
+datapath = '/'.join(os.path.abspath(sys.argv[0]).split('/')[0:-2])+'/_data'
```
Cleaner: 
```suggestion
ownpath = os.path.abspath(sys.argv[0])
repopath = os.path.dirname(os.path.dirname(ownpath))
datapath = repopath +'/_data'
```
... possibly for a future update

fingolfin commented on etc/update-dates.py:
```diff
@@ -1,37 +1,49 @@
+if os.getcwd().split('/')[-1] == 'etc':
+    ogfile = yaml.safe_load(open("../_data/examples.yml"))
+else:
+    ogfile = yaml.safe_load(open("_data/examples.yml"))
```
like in the other file this hack should be replaced

fingolfin commented:
Since website updates are basically completely broken right now, we need to move quickly with this.

I'd be OK with merging this now and fixing it later, but as long as we set no `API_KEY` it won't do much. But perhaps that good...

But if this fails to work again, I'd rather rip it all out for now (well, deactivate it in the `update.sh` script).


Overall I'd feel much better if I knew these scripts were tested at all; either by you, locally, or (much better on the long run) in a GitHub Action -- so that e.g. syntax errors or typos in variables names are caught before we try it in the wild. But this can come later...

fingolfin committed:
Merge branch 'gh-pages' into ak96/rewrite-in-python-attempt-2

fingolfin committed:
update

fingolfin committed:
more

aaruni96 commented:
> but as long as we set no API_KEY it won't do much

I set the API_KEY in the systemd unit file (mentioned at the top of this PR). I will also add this to the readme.
